### PR TITLE
[AsmParser] Don't require value numbers to be consecutive

### PR DESCRIPTION
**Pull request by nikic**
Currently, the IR parser requires that %n style numbered values are consecutive. This means that the IR becomes invalid as soon as you remove an instruction, argument or block. This makes it very annoying to modify IR without running it through instnamer first.

I don't think there is any good reason to impose this requirement. This PR relaxes it to allow value IDs to be non-consecutive, but it still keeps the requirement that they're increasing (i.e. you can't skip a value number and then assign it later).

I've kept the underlying representation the same and added a guard against skipping large ranges to guard against degenerate cases where the NumberedVals vector would become huge. If people prefer, I could change NumberedVals from a vector to a map + a counter instead, in which case we could support arbitrary numbers without large increases in memory usage.

This only implements support for skipping numbers for local values. We should extend this to global values in the future as well.

### CODE REVIEW

File: llvm/test/Assembler/call-nonzero-program-addrspace-2.ll [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 ; Check that numbered variables in a nonzero program address space 200 can be used in a call instruction
 
-define i8 @test_unnamed(ptr, ptr addrspace(42) %0) {
+define i8 @test_unnamed(ptr, ptr addrspace(42) %1) {
```
Comment (nikic) [inline llvm/test/Assembler/call-nonzero-program-addrspace-2.ll:6]: This is a bug in the previous implementation: While `%0` is written here, it was actually interpreted as `%1`.

**Comment by nikic** [comment]
> I've kept the underlying representation the same and added a guard against skipping large ranges to guard against degenerate cases where the NumberedVals vector would become huge. If people prefer, I could change NumberedVals from a vector to a map + a counter instead, in which case we could support arbitrary numbers without large increases in memory usage.

I went ahead and did this. The restriction on value skipping seemed pretty odd, and I don't think we really care about the performance difference between using a vector and a map when it comes to IR parsing.

**Comment by dtcxzyw** [comment]
> I don't think there is any good reason to impose this requirement. This PR relaxes it to allow value IDs to be non-consecutive, but it still keeps the requirement that they're increasing (i.e. you can't skip a value number and then assign it later).

https://github.com/llvm/llvm-project/blob/2db9244b6f66a4fd4c937de7b33a5b6dda4a06a8/llvm/docs/LangRef.rst#L131-L135

Could you please update the LangRef as well?

**Review by nhaehnle — APPROVED** [review]
Nice, I've run into this issue in the past and it was always slightly annoying to have to go back and get out the instnamer. LGTM.

Side note: the duplicate argument parsing in `parseArgumentList` is really awkward.

**Pull request merged**